tinygrad/codegen/linearizer.py

```diff
@@ -293,14 +291,23 @@ def upcast_strides(buf:int):
       loaded_buffers[self.bufs[-1]] = self.global_load(-1, fake_global_idxs+local_idxs+fake_reduce_idxs+upcast_idxs, barrier=barrier)
 
       # there's no AST here (and there's no shape for the reduce LazyOp)
-      self.ast_parse(LazyOp(reduceop.op, (LazyOp(BufferOps.LOAD, (), self.bufs[-1]),)), acc, self.acc_offsets(-1), loaded_buffers, do_reduce=True, loop_ctx=loop_ctx)  # noqa: E501
+      self.ast_parse(LazyOp(reduceop.op, (LazyOp(BufferOps.LOAD, (), self.bufs[-1]),)), acc, self.acc_offsets(-1), loaded_buffers, do_reduce=True, loop_ctx=loop_ctx, if_ctx=barrier)  # noqa: E501
 
       # end the late reduce loop
       self.load_cache.clear()
 
-      # all local indices which were used for group_for_reduce are not valid any more and should be replaced with fake NumNode(0), since they have
-      # been rewritten with fake end_local_idxs.
-    return (acc, loaded_buffers, fake_reduce_idxs, local_idxs[:self.local_dims] + [NumNode(0) for i in range(self.group_for_reduces)], upcast_idxs)
+      # if there are more reduceops, store the final value into the temp buffer, the load it back into all threads
+      if len(self.reduce_acc) < len(self.reduceops)-1:
+        for j in self.upcast_in_mid_reduce_axes:
+          self.upcasted -= 1
+          self.group_for_reduces += 1
+        reduce_buf_uop = self.buf_uops[-1]
+        assert reduce_buf_uop is not None, "Local reduce buf must have been uoped at this point"
+        stores = self.global_store(-1, [NumNode(0)]*len(self.sts[-1].shape), acc)
+        endif = self.uops.add(UOps.ENDIF, None, (barrier,), cachable=False)
+        barrier = self.uops.add(UOps.BARRIER, None, (endif, *stores), cachable=False)
+        acc = self.global_load(-1, [NumNode(0)]*len(self.sts[-1].shape), barrier=barrier)
+    return acc
```

Comment:
this will store the result of the local reduce and load it back into each thread

Reply:
why is the ENDIF still there?